api/hypershift/v1alpha1/nodepool_types.go

```diff
@@ -888,16 +888,55 @@ type AgentNodePoolPlatform struct {
 	AgentLabelSelector *metav1.LabelSelector `json:"agentLabelSelector,omitempty"`
 }
 
-type AzureNodePoolPlatform struct {
-	VMSize string `json:"vmsize"`
-	// ImageID is the id of the image to boot from. If unset, the default image at the location below will be used:
-	// subscription/$subscriptionID/resourceGroups/$resourceGroupName/providers/Microsoft.Compute/images/rhcos.x86_64.vhd
+type AzureVMImageType string
+
+const (
+	ImageID          AzureVMImageType = "ImageID"
+	AzureMarketplace AzureVMImageType = "AzureMarketplace"
+)
+
+// AzureVMImage represents the different types of image data that can be provided for an Azure VM.
+// +union
+type AzureVMImage struct {
+	// Type is the type of image data that will be provided to the Azure VM. This can be either "ImageID" or
+	// "MarketplaceImageInfo".
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=ImageID;AzureMarketplace
+	// +unionDiscriminator
+	Type AzureVMImageType `json:"azureImageType"`
+
+	// ImageID is the Azure resource ID of a VHD image to use to boot the Azure VMs from.
+	//
 	// +optional
+	// +unionMember, optional
 	ImageID string `json:"imageID,omitempty"`
+
+	// MarketplaceImageInfo contains the Azure Marketplace image info to use to boot the Azure VMs from.
+	//
+	// +optional
+	// +unionMember, optional
```

Comment:
All union member fields must be `+optional` since they become required (you'll need CEL for this for now) when the discriminator matches their value.

The `+unionMember,optional`, means that this extra info is optional even when the discriminator points to it.

It's not typically used that often, but I suspect looking at this API, you don't want optional members.

You can see examples here to help, https://github.com/openshift/api/blob/465787efd0d69f12beaad9410a3812d823e78c2b/example/v1/types_stable.go#L108-L125

Reply:
Thanks for the info! Yeah I don't think we want the optional part in `+unionMember,optional` given the statement in the example Joel provided `requiredMember is required when type is RequiredMember, and forbidden otherwise`

Reply:
This looks right to me but I keep getting an error when I try to apply the CR. Any suggestions?
```
// AzureVMImageType is a union discriminator for AzureVMImage
// +kubebuilder:validation:Enum:=ImageID;AzureMarketplace
type AzureVMImageType string

const (
 ImageID          AzureVMImageType = "ImageID"
 AzureMarketplace AzureVMImageType = "AzureMarketplace"
)

// AzureVMImage represents the different types of image data that can be provided for an Azure VM.
// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type =='ImageID' ? has(self.imageID) : !has(self.imageID)",message="ImageID is required when type is ImageID, and forbidden otherwise"
// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'AzureMarketplace' ? has(self.azureMarketplace) : !has(self.azureMarketplace)",message="AzureMarketplace is required when type is AzureMarketplace, and forbidden otherwise"
// +union
type AzureVMImage struct {
 // Type is the type of image data that will be provided to the Azure VM. This can be either "ImageID" or
 // "AzureMarketplace".
 //
 // +kubebuilder:validation:Required
 // +unionDiscriminator
 Type AzureVMImageType `json:"azureImageType"`

 // ImageID is the Azure resource ID of a VHD image to use to boot the Azure VMs from.
 //
 // +optional
 // +unionMember
 ImageID *string `json:"imageID,omitempty"`

 // AzureMarketplace contains the Azure Marketplace image info to use to boot the Azure VMs from.
 //
 // +optional
 // +unionMember
 AzureMarketplace *MarketplaceImage `json:"azureMarketplace,omitempty"`
}
```

The error message
```
Error: CustomResourceDefinition.apiextensions.k8s.io "nodepools.hypershift.openshift.io" is invalid: [spec.versions[1].schema.openAPIV3Schema.properties[spec].properties[platform].properties[azure].properties[image].allOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural, spec.versions[1].schema.openAPIV3Schema.properties[spec].properties[platform].properties[azure].properties[image].allOf[1].x-kubernetes-validations: Forbidden: must be empty to be structural]
CustomResourceDefinition.apiextensions.k8s.io "nodepools.hypershift.openshift.io" is invalid: [spec.versions[1].schema.openAPIV3Schema.properties[spec].properties[platform].properties[azure].properties[image].allOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural, spec.versions[1].schema.openAPIV3Schema.properties[spec].properties[platform].properties[azure].properties[image].allOf[1].x-kubernetes-validations: Forbidden: must be empty to be structural]
```

Reply:
Can you either push the changes or show the part of the generated openAPI schema that it's referring to?

Reply:
I pushed the changes in the latest commit - https://github.com/openshift/hypershift/pull/4191/commits/bcbffa511c8da47d16be3c421d6b18c640957490

Reply:
Planning to follow up on this in a separate PR - https://github.com/openshift/hypershift/pull/4460